redis/src/test/scala/zio/redis/ScriptingSpec.scala

```diff
@@ -0,0 +1,260 @@
+            sha <- scriptLoad(lua)
+            res <- evalSha(sha, Chunk(key), Chunk(arg)).returning[String].either
+          } yield assert(res)(isLeft(isSubtype[ProtocolError](hasField("message", _.message, equalTo(customError)))))
+        },
+        testM("throw custom error from script") {
+          import ScriptingSpec.simpleStringOutput
+          for {
+            key    <- uuid
+            arg    <- uuid
+            myError = "My Error"
+            lua     = s"""return redis.error_reply("${myError}")"""
+            sha    <- scriptLoad(lua)
+            res    <- evalSha(sha, Chunk(key), Chunk(arg)).returning[String].either
+          } yield assert(res)(isLeft(isSubtype[ProtocolError](hasField("message", _.message, equalTo(myError)))))
+        },
+        testM("throw NoScript error if script isn't found in cache") {
+          import ScriptingSpec.simpleStringOutput
+          val lua                       = """return "1""""
+          val error                     = "No matching script. Please use EVAL."
+          val emptyInput: Chunk[String] = Chunk.empty
+          for {
+            res <- evalSha(lua, emptyInput, emptyInput).returning[String].either
+          } yield assert(res)(isLeft(isSubtype[NoScript](hasField("message", _.message, equalTo(error)))))
+        }
+      ),
+      suite("scriptExists")(
+        testM("return true if scripts are found in the cache") {
+          val lua1 = """return "1""""
+          val lua2 = """return "2""""
+          for {
+            sha1 <- scriptLoad(lua1)
+            sha2 <- scriptLoad(lua2)
+            res  <- scriptExists(sha1, sha2)
+          } yield assert(res)(equalTo(Chunk(true, true)))
+        },
+        testM("return false if scripts aren't found in the cache") {
+          val lua1 = """return "1""""
+          val lua2 = """return "2""""
+          for {
+            res <- scriptExists(lua1, lua2)
+          } yield assert(res)(equalTo(Chunk(false, false)))
+        }
+      ),
+      suite("scriptLoad")(
+        testM("return OK") {
+          val lua = """return "1""""
+          for {
+            sha <- scriptLoad(lua)
+          } yield assert(sha)(isSubtype[String](anything))
+        },
+        testM("throw an error when incorrect script was sent") {
+          val lua   = ";"
+          val error = "Error compiling script (new function): user_script:1: unexpected symbol near ';'"
+          for {
+            sha <- scriptLoad(lua).either
+          } yield assert(sha)(isLeft(isSubtype[ProtocolError](hasField("message", _.message, equalTo(error)))))
+        }
+      )
+    )
+}
+
+object ScriptingSpec {
+
+  final case class CustomInputValue(name: String, age: Long)
+
+  object CustomInputValue {
+    implicit val encoder: Input[CustomInputValue] = Input.Tuple2(StringInput, LongInput).contramap { civ =>
+      (civ.name, civ.age)
+    }
+  }
+
+  final case class CustomData(count: Long, avg: Long, pair: (Int, String))
+
+  object CustomData {
```

Comment:
I would not nest items not directly related to CustomData in its companion object.